src/transaction/log_page_buffer.c

```diff
@@ -2141,30 +1640,63 @@ logpb_copy_page (THREAD_ENTRY * thread_p, LOG_PAGEID pageid, LOG_CS_ACCESS_MODE
       log_csect_entered = true;
     }
 
-  /* TODO: Can we use a latch free structure here? */
-  START_EXCLUSIVE_ACCESS_LOG_PB (rv, thread_p);
-
-  log_bufptr = (LOG_BUFFER *) mht_get (log_Pb.ht, &pageid);
-  if (log_bufptr != NULL)
+  if (pageid == LOGPB_HEADER_PAGE_ID)
     {
-      memcpy (log_pgptr, &log_bufptr->logpage, LOG_PAGESIZE);
-      ret_pgptr = log_pgptr;
+      /* copy header page into log_pgptr */
+      log_bufptr = &log_Pb.header_buffer;
+      if (log_bufptr->pageid == NULL_PAGEID)
+	{
+	  rv = logpb_read_page_from_file (thread_p, pageid, access_mode, log_pgptr);
+	  if (rv != NO_ERROR || log_pgptr == NULL)
+	    {
+	      rv = ER_FAILED;
+	      goto exit;
+	    }
+	  mnt_log_fetch_ioreads (thread_p);
+	  stat_page_found = PERF_PAGE_MODE_OLD_LOCK_WAIT;
+	}
+      else
+	{
+	  memcpy (log_pgptr, log_bufptr->logpage, LOG_PAGESIZE);
+	}
+
+      goto exit;
     }
 
-  END_EXCLUSIVE_ACCESS_LOG_PB (rv, thread_p);
+  log_bufptr = &log_Pb.buffers[logpb_get_log_buffer_index ((int) pageid)];
+  if (log_bufptr->pageid == pageid)
+    {
+      /* Copy page from log_bufptr into log_pgptr */
+      memcpy (log_pgptr, log_bufptr->logpage, LOG_PAGESIZE);
 
-  if (log_bufptr == NULL)
+      if (log_bufptr->pageid == pageid)
+	{
+	  goto exit;
+	}
+    }
+
+  /* Could not get from log page buffer cache */
+  rv = logpb_read_page_from_file (thread_p, pageid, access_mode, log_pgptr);
+  if (rv != NO_ERROR || log_pgptr == NULL)
+    {
+      rv = ER_FAILED;
+      goto exit;
+    }
+  mnt_log_fetch_ioreads (thread_p);
+  stat_page_found = PERF_PAGE_MODE_OLD_LOCK_WAIT;
+  if (log_pgptr == NULL)
```

Comment:
The race was resolved here. I want the "if" block, and this is exactly why, because I want to check if the buffer is not modified. I will put this if back, I forgot why was there. With this condition, there will be no race problems.